component/prometheus/fanout.go

```diff
@@ -0,0 +1,101 @@
+package prometheus
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/prometheus/prometheus/model/exemplar"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/metadata"
+
+	"github.com/prometheus/prometheus/storage"
+)
+
+var _ storage.Appendable = (*Fanout)(nil)
+
+// Fanout supports the default Flow style of appendables since it can go to multiple outputs. It also allows the intercepting of appends.
+type Fanout struct {
+	mut sync.RWMutex
+	// children is where to fan out.
+	children []storage.Appendable
+	// ComponentID is what component this belongs to.
+	componentID string
+}
+
+// NewFanout creates a fanout appendable.
+func NewFanout(children []storage.Appendable, componentID string) *Fanout {
+	return &Fanout{
+		children:    children,
+		componentID: componentID,
+	}
+}
+
+// UpdateChildren allows changing of the children of the fanout.
+func (f *Fanout) UpdateChildren(children []storage.Appendable) {
+	f.mut.Lock()
+	defer f.mut.Unlock()
+	f.children = children
+}
+
+// Appender satisfies the Appendable interface.
+func (f *Fanout) Appender(ctx context.Context) storage.Appender {
+	f.mut.RLock()
+	defer f.mut.RUnlock()
+
+	app := &appender{
+		children:    make([]storage.Appender, 0),
+		componentID: f.componentID,
+	}
+	for _, x := range f.children {
+		if x == nil {
+			continue
+		}
+		app.children = append(app.children, x.Appender(ctx))
+	}
+	return app
+}
+
+var _ storage.Appender = (*appender)(nil)
+
+type appender struct {
+	children    []storage.Appender
+	componentID string
+}
+
+// Append satisfies the Appender interface.
+func (a *appender) Append(ref storage.SeriesRef, l labels.Labels, t int64, v float64) (storage.SeriesRef, error) {
+	if ref == 0 {
+		ref = storage.SeriesRef(GlobalRefMapping.GetOrAddGlobalRefID(l))
+	}
+	for _, x := range a.children {
+		_, _ = x.Append(ref, l, t, v)
+	}
+	return ref, nil
+}
+
+// Commit satisfies the Appender interface.
+func (a *appender) Commit() error {
+	for _, x := range a.children {
+		_ = x.Commit()
+	}
+	return nil
+}
+
+// Rollback satisifies the Appender interface.
+func (a *appender) Rollback() error {
+	for _, x := range a.children {
+		_, _ = x, a.Rollback()
+	}
+	return nil
+}
```

Comment:
Have you considered using a multierror here and returning the error back to the caller so it can make the decision for what to do with error handling?

Ultimately I don't think the fanout should be hiding errors and it should be the callers responsibility.